### PR TITLE
feat: Add `filterNode` option to `prettyDOM`

### PR DESCRIPTION
**Pull request by eps1lon**
BREAKING CHANGE: `<script />`, `<style />` and comment nodes are now ignored by default in `prettyDOM` .If you whish to return to the old behavior, use a custom `filterNode` function. In this case `prettyDOM(element, { filterNode: () => true })`.

I would find this change useful. But I believe this is possible to do in userland so I'm preparing a PR that just implements this in userland. We can then decide whether we want to defer to userland, release it as a breaking change or make it backwards compatible.

**What**:

Ignore `<script />`, `<style />` and comment nodes in `prettyDOM`

**Why**:

test frameworks like `karma` and `mocha` add `<script />` and comment nodes that [add quite a bit of noise making debugging harder](https://app.circleci.com/pipelines/github/mui-org/material-ui/38685/workflows/83279e68-0b89-4683-8c07-748ecf9cfc64/jobs/228457/parallel-runs/0/steps/0-112).

Would also improve https://github.com/testing-library/dom-testing-library/issues/902

**How**:

Fork `DOMElement` plugin from `pretty-dom` to allow filtering single `Node` instances.

**Checklist**:

- [ ] Documentation added to the
      [docs site](https://github.com/testing-library/testing-library-docs) 
      Will do when we release the stable v8
- [x] Tests
- [x] Typescript definitions updated
- [x] Ready to be merged

### CODE REVIEW

**Comment by MatanBobi** [comment]
I can see the value of this one but it is a breaking change and should be considered thoroughly. 
What about adding this one behind a configuration?

**Comment by eps1lon** [comment]
> I can see the value of this one but it is a breaking change and should be considered thoroughly.
> What about adding this one behind a configuration?

I think I referred to all of this in the PR descript: "We can then decide whether we want to defer to userland, release it as a breaking change or make it backwards compatible."

**Comment by nickmccurdy** [comment]
Perhaps we could hide this behind a config boolean, and set the default to true on a major release.

**Comment by eps1lon** [comment]
I appreciate the review but let's focus on the high level problem: Do we want this as a feature or breaking change? What is the use case for the current behavior?

Then we can talk about the implementation. The implementation is problematic for a feature since it adds configuration but trivial to do.

**Comment by MatanBobi** [comment]
> I appreciate the review but let's focus on the high level problem: Do we want this as a feature or breaking change? What is the use case for the current behavior?

I'm actually not so sure I'm in favor of adding this one. I think that isn't a part of our responsibility and may just cause bugs/issues in some cases. As you already wrote, this can be patched on the user-land so I'm not sure a solution on our end is really necessary.
That's just my opinion of course.

**Comment by eps1lon** [comment]
> I think that isn't a part of our responsibility and may just cause bugs/issues in some cases.

I think neither of these statements are relevant. What is our responsibility is better error diagnostics. So the question is if this helps or hinders error diagnostic which we should talk about.

Causing bugs is inherent to any change. The question is how likely they are.

> As you already wrote, this can be patched on the user-land so I'm not sure a solution on our end is really necessary.

I **believe**. I haven't actually done it so let's not jump to conclusions and assume that the userland implementation is trivial.

**Comment by kentcdodds** [comment]
We already have `DEFAULT_IGNORE_TAGS`: https://github.com/testing-library/dom-testing-library/blob/ffc8f26d48543de95c7fff4e5d961cc4beaa3872/src/config.ts#L44

That's used for the `ignore` option on queries. I wonder whether it'd be possible to ignore everything that the failing query would ignore by default. That would be more correct anyway because right now we're displaying stuff that actually couldn't be queried.

**Comment by kentcdodds** [comment]
To be clear, I'm in favor of this change in light of that :)

**Comment by eps1lon** [comment]
> We already have DEFAULT_IGNORE_TAGS:

That makes a lot of sense. I was a bit suprised that I couldn't find existing implementation. Will try to see if I can leverage `DEFAULT_IGNORE_TAGS` usage instead.

**Comment by kentcdodds** [comment]
To be clear, I think we should try to use whatever the `ignore` configuration is for the query that's failed. In most cases this will be the default, but it's an important distinction.

**Comment by kentcdodds** [comment]
In fact, it may be useful to also display the `ignore` configuration as well so folks know that not everything is being displayed.

File: src/shared.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export const DEFAULT_IGNORE_TAGS = 'script, style'
```
Comment (eps1lon) [inline src/shared.ts:1]: Breaks up a circular dep between `config` and `prettyDOM`

File: src/DOMElementFilter.ts [hunk 2]
```diff
@@ -0,0 +1,259 @@
+/**
+ * Source: https://github.com/facebook/jest/blob/master/packages/pretty-format/src/plugins/DOMElement.ts
```
Comment (eps1lon) [inline src/DOMElementFilter.ts:2]:
```suggestion
 * Source: https://github.com/facebook/jest/blob/e7bb6a1e26ffab90611b2593912df15b69315611/packages/pretty-format/src/plugins/DOMElement.ts
```
Permalink

File: src/DOMElementFilter.ts [hunk 3]
```diff
@@ -0,0 +1,259 @@
+/**
+ * Source: https://github.com/facebook/jest/blob/e7bb6a1e26ffab90611b2593912df15b69315611/packages/pretty-format/src/plugins/DOMElement.ts
+ */
+import type {Config, NewPlugin, Printer, Refs} from 'pretty-format'
```
Comment (eps1lon) [inline src/DOMElementFilter.ts:4]: Requires https://github.com/facebook/jest/pull/11342

Might need to use module augmentation depending on the targetted release of https://github.com/facebook/jest/pull/11342

File: src/config.ts [hunk 4]
```diff
@@ -31,8 +31,18 @@ let config: InternalConfig = {
 
   // called when getBy* queries fail. (message, container) => Error
   getElementError(message, container) {
+    const prettifiedDOM = prettyDOM(container)
     const error = new Error(
-      [message, prettyDOM(container)].filter(Boolean).join('\n\n'),
+      [
+        message,
+        prettifiedDOM.length > 0
+          ? `Ignored nodes: comments, <script />, <style />\n${prettyDOM(
```
Comment (kentcdodds) [inline src/config.ts:39]: I'm not sure I'm a fan of this note.

**Review by kentcdodds — COMMENTED** [review]
Thanks for this work!

Not excited about vendoring the DOMElementFilter file. I'm guessing they weren't receptive to a PR for this feature?

How much work would it be to take the `ignore` option into account for the errors that are thrown? For example, let's say I'm trying to find text within a style tag. I would set the `selector` and `ignore` values like so:

```tsx
screen.getByText('.css-selector', {selector: 'style', ignore: 'script'})
```

Then my error should only show the things that could have matched. And it should show the selector and ignore configuration options as well.

So would it be possible to automatically filter the error message print out in a way that matches the `selector` and `ignore` values? I just think that a mismatch in those two could lead to confusing results.

I think it would amount to passing the `selector` and `ignore` options to `prettyDOM` and when those are provided, a `filterNode` function is created which only let's nodes that match the selector and don't match the `ignore` through. That way the error message displayed shows *exactly* what the user could have queried for given the options.

**Comment by eps1lon** [comment]
> Not excited about vendoring the DOMElementFilter file. 

Me neither :smile: 

>  I'm guessing they weren't receptive to a PR for this feature?

It was pretty cumbersome to integrate while not breaking all the other tests. They generally weren't receptive to filtering stuff out at all which makes sense considering `pretty-format` is used for generating snapshot matchers. And if you want to filter stuff out from snapshots you're better off implementing custom snapshot matchers. But that's not how we're using `pretty-format`. 

I think I'm generally fine with the "fork now, upstream later" approach here.

> I just think that a mismatch in those two could lead to confusing results.

This is a real concern I agree. But we might want to revisit if we want to consider `screen.getByText('.css-selector', {selector: 'style', ignore: 'script'})` valid in the first place. I can't imagine that this is a useful query. It looks like querying implementation details when you're better off using `querySelect(.css-selector)` and assert the computed styles of that element.
In other words: Is the effort worth the edge cases it covers or does this edge case virtually never happen i.e. we're aware of the defect and are willing to push a bug fix if somebody needs it.

**Comment by kentcdodds** [comment]
> I think I'm generally fine with the "fork now, upstream later" approach here.

Agreed.

> valid in the first place.

Yeah, I was worried my contrived example would be distracting 😅

I think in general the biggest problem is the mismatch between what we see in the error and what we ignore/select. If you don't use `ignore` or `select` then it's fine because we can just go with the defaults, but once you use those options then it can become pretty unhelpful, and that's what I would like to avoid.

> Is the effort worth the edge cases it covers or does this edge case virtually never happen i.e. we're aware of the defect and are willing to push a bug fix if somebody needs it.

That's fair. I currently don't have a use case for `ignore` or `selector` personally so I would be satisfied with a TODO code comment just so folks know that we'd like to support this in the future (rather than trying to figure out why we didn't do it in the first place).

**Comment by eps1lon** [comment]
> I currently don't have a use case for ignore or selector personally so I would be satisfied with a TODO code comment just so folks know that we'd like to support this in the future (rather than trying to figure out why we didn't do it in the first place).

Sounds good. I just go ahead and add a test so that it's really "just" about fixing the problem. Makes it easier to get started contributing.

File: types/pretty-dom.d.ts [hunk 5]
```diff
@@ -1,13 +1,21 @@
 import * as prettyFormat from 'pretty-format'
 
+export interface PrettyFormatOptions extends prettyFormat.OptionsReceived {
+  /**
+   * Given a `Node` return `false` if you whish to ignore that node in the output.
```
Comment (timdeschryver) [inline types/pretty-dom.d.ts:5]:
```suggestion
   * Given a `Node` return `false` if you wish to ignore that node in the output.
```

**Review by timdeschryver — COMMENTED** [review]
This would be a great addition to ATL, where I believe that Karma is the most popular. Just a small typo...

File: types/pretty-dom.d.ts (hunk 5 above)
Comment (eps1lon) [reply types/pretty-dom.d.ts:5]: English language: which, wish :man_facepalming:

**Review by kentcdodds — APPROVED** [review]
Solid 👍 Thanks!
